tests/Compiler/Type/PhpDocTypeUtilsTest.php

```diff
@@ -1036,4 +1023,86 @@ private static function provideIsSubTypeOfDataInner(): iterable
         ];
     }
 
+    #[DataProvider('provideInferGenericParameterData')]
+    public function testInferGenericParameter(
+        string $type,
+        string $genericTypeName,
+        int $parameter,
+        string $expectedResult
+    ): void
+    {
+        self::assertEquals(
+            $this->parseType($expectedResult),
+            PhpDocTypeUtils::inferGenericParameter($this->parseType($type), $genericTypeName, $parameter),
+        );
+    }
+
+    /**
+     * @return iterable<array{0: string, 1: string, 2: int, 3: string}>
+     */
+    public static function provideInferGenericParameterData(): iterable
+    {
+        yield [
+            'list<int>',
+            'list',
+            0,
+            'int',
+        ];
+
+        yield [
+            'list<int>|list<string>',
+            'list',
+            0,
+            'int|string',
+        ];
+
+        yield [
+            'array<int>',
+            'array',
+            0,
+            'mixed',
+        ];
+
+        yield [
+            'array<int>',
+            'array',
+            1,
+            'int',
+        ];
+
+        yield [
+            'list<string>',
+            'array',
+            0,
+            'int',
+        ];
+
+        yield [
+            'list<string>',
+            'array',
+            1,
+            'string',
+        ];
+
+        yield [
+            'ShipMonk\InputMapper\Runtime\Optional<Countable> & ShipMonk\InputMapper\Runtime\Optional<Traversable>',
+            'ShipMonk\InputMapper\Runtime\Optional',
+            0,
+            'Countable & Traversable',
+        ];
```

Comment:
Can you add testcase with something similar to this? https://github.com/phpstan/phpstan/issues/9198

Reply:
I'm not sure what you mean. There is case for union.

```php
        yield [
            'list<int>|list<string>',
            'list',
            0,
            'int|string',
        ];
```

There is no normalization for unions for now. But it you ask what is the value type of `list<int>|list<string>`, the answer has to be `int|string` which is an information loss. To avoid loosing the information, you must never pass union to `inferGenericParameter` and handle the union on the caller side. But that is not always pratical.